src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.ClassifiedLineCache.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Classification;
+
+internal partial class SyntacticClassificationTaggerProvider
+{
+    /// <summary>
+    /// it is a helper class that encapsulates logic on holding onto last classification result.  This type has
+    /// thread affinity on the UI thread, so it doesn't need any synchronization.
+    /// </summary>
+    /// <remarks>
+    /// Empirical testing shows that when paging through a file, 25% of syntactic classification requests can be
+    /// returned from a simple LRU cache.  When arrowing up and down in a file, the cache hit rate is 85%.  This can
+    /// substantially speed up these requests, especially in large files, as it means we can avoid walking syntax
+    /// trees and reclassifying lines we just classified.
+    /// </remarks>
+    private sealed class ClassifiedLineCache(IThreadingContext threadingContext)
+    {
+        /// <summary>
+        /// Ensure that we don't cache incredibly long lines (for example, a minified JavaScript file).  This also
+        /// ensures if we were asked by some party to classify a large section of the file, that we don't attempt to
+        /// store that.  The primary purpose of this cache is for the common case of returning cached lines when the
+        /// editor calls into us to classify them.
+        /// </summary>
+        private const int MaxClassificationsCount = 1024;
+
+        /// <summary>
+        /// This number was picked as a reasonable number of lines to classify given regular screen sized.  This is
+        /// about 6x larger than a normal number of lines in VS at standard resolutions, and handles the case of
+        /// users with vertical monitors and small font/zoom settings..
+        /// </summary>
+        private const int CacheSize = 256;
+
+        private readonly IThreadingContext _threadingContext = threadingContext;
+
+        // Mutating state.  No need for locks as we only execute on the UI thread (and throw if we're not on that thread).
+
+        private ITextSnapshot? _snapshot;
+
+        private readonly LinkedList<Span> _spans = [];
+        private readonly Dictionary<Span, (LinkedListNode<Span> node, SegmentedList<ClassifiedSpan> classifiedSpans)> _spanToClassifiedSpansMap = [];
```

Comment:
>SegmentedList [](http://example.com/codeflow?start=70&length=13)

would it make sense for this to be in the LinkedListNode instead?

Reply:
indeed. much simpler.